rust/syntect-plugin/src/stackmap.rs

```diff
@@ -0,0 +1,120 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! StackMap is a simple nested map type (a trie) used to map `StackScope`s to
+//! u32s so they can be efficiently sent to xi-core.
+//!
+//! For discussion of this approach, see [this
+//! issue](https://github.com/google/xi-editor/issues/284).
+
+use std::collections::HashMap;
+
+use syntect::parsing::Scope;
+
+
+#[derive(Debug, Default)]
+struct Node {
+    value: Option<u32>,
+    children: HashMap<Scope, Node>,
+}
+
+#[derive(Debug, Default)]
+/// Nested lookup table for stacks of scopes.
+pub struct StackMap {
+    next_id: u32,
+    scopes: Node,
+}
+
+#[derive(Debug, PartialEq)]
+/// Result type for `StackMap` lookups. Used to communicate to the user
+/// whether or not a new identifier has been assigned, which will need to
+/// be communicated to the peer.
+pub enum LookupResult {
+    Existing(u32),
+    New(u32),
+}
+
+impl Node {
+    pub fn new(value: u32) -> Self {
+        Node { value: Some(value), children: HashMap::new() }
+    }
+
+    fn get_value(&mut self, stack: &[Scope], next_id: u32) -> LookupResult {
+        // if this is last item on the stack, get the value, inserting if necessary.
+        let first = stack.first().unwrap();
+        if stack.len() == 1 {
+            if !self.children.contains_key(first) {
+                self.children.insert(first.to_owned(), Node::new(next_id));
+                return LookupResult::New(next_id)
+            }
+
+            // if key exists, value still might not be assigned:
+            let needs_value = self.children.get(first).unwrap().value.is_none();
+            if needs_value {
+                let mut node = self.children.get_mut(first).unwrap();
+                node.value = Some(next_id);
+                return LookupResult::New(next_id)
+            } else {
+               let value = self.children.get(first).unwrap().value.unwrap();
+                return LookupResult::Existing(value)
+            }
+        }
+        // not the last item: recurse, creating node as necessary
+        if self.children.get(first).is_none() {
+            self.children.insert(first.to_owned(), Node::default());
+        }
+        self.children.get_mut(first).unwrap().get_value(&stack[1..], next_id)
+    }
+}
+
+impl StackMap {
+    /// Returns the identifier for this stack, creating it if needed.
+    pub fn get_value(&mut self, stack: &[Scope]) -> LookupResult {
+        assert!(!stack.is_empty());
+        let result = self.scopes.get_value(stack, self.next_id);
+        if result.is_new() {
```

Comment:
This could become `if let LookupResult::New(_) = result {` and then the method below would be unnecessary.

But I do sympathize in that I wish Rust had an easier way to get "is this thing this type of the enum" as a boolean by default.